core/src/main/java/org/frankframework/pipes/ReplacerPipe.java

```diff
@@ -70,24 +95,77 @@ public void configure() throws ConfigurationException {
 
 	@Override
 	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
-		ReplacingInputStream inputStream = getReplacingInputStream(message, getFind(), getReplace());
+		try {
+			// Create a ReplacingInputStream for find/replace
+			ReplacingInputStream replacingInputStream = new ReplacingInputStream(message.asInputStream(), find, replace, isReplaceNonXmlChars(),
+					getNonXmlReplacementCharacter(), isAllowUnicodeSupplementaryCharacters()
+			);
 
-		Message result = new Message(inputStream, message.copyContext().withoutSize());
-		// As we wrap the input-stream, we should make sure it's not closed when the session is closed as that might close this stream before reading it.
-		message.unscheduleFromCloseOnExitOf(session);
-		result.closeOnCloseOf(session, this);
+			// Replace parameters
+			ReplacingVariablesInputStream replaceParametersStream = new ReplacingVariablesInputStream(replacingInputStream, "?",
+					getKeyValueMapForParameters(message, session)
+			);
 
-		return new PipeRunResult(getSuccessForward(), result);
-	}
+			// Wrap for 'substitute vars' if necessary
+			ReplacingVariablesInputStream inputStream = wrapWithSubstituteVarsInputStreamIfNeeded(session, replaceParametersStream);
 
-	private ReplacingInputStream getReplacingInputStream(Message message, String find, String replace) throws PipeRunException {
-		try {
-			return new ReplacingInputStream(message.asInputStream(), find, replace, isReplaceNonXmlChars(), getNonXmlReplacementCharacter(), isAllowUnicodeSupplementaryCharacters());
+			Message result = new Message(inputStream);
+
+			return new PipeRunResult(getSuccessForward(), result);
 		} catch (IOException e) {
 			throw new PipeRunException(this, "cannot open stream", e);
 		}
 	}
 
+	/**
+	 * If subsituteVars is true, we need to wrap the inputStream again to substitute ${} syntax variables with
+	 * system properties, session variables and application properties.
+	 *
+	 * @param session
+	 * @param replaceParametersStream
+	 * @return
+	 */
+	private ReplacingVariablesInputStream wrapWithSubstituteVarsInputStreamIfNeeded(PipeLineSession session,
+																					ReplacingVariablesInputStream replaceParametersStream) {
+		if (substituteVars) {
+			Map<String, String> sessionValues = session.entrySet().stream()
+					.collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().toString()));
+
+			return new ReplacingVariablesInputStream(replaceParametersStream, "$",
+					getKeyValueMapForParametersAndAppConstants(sessionValues)
+			);
+		}
+
+		return replaceParametersStream;
+	}
+
+	/**
+	 * @return the appConstants as a key/value map combined with the pipe's parameters
+	 */
+	private Map<String, String> getKeyValueMapForParametersAndAppConstants(Map<String, String> keyValueMapForParameters) {
+		Map<String, String> parametersAndConstants = appConstants.entrySet().stream()
```

Comment:
Deze variable heet `parametersAndConstants` maar ik zie alleen AppConstants?

Reply:
Op regel 153 worden ook de parameters eraan toegevoegd

Reply:
Ok maar dan haal je wel 2 dingen door elkaar.
`${properties}` gaan via AppConstants (en alleen AppConstants).
`?{parameters}` gaan via params (en dus alleen via Parameters).

Reply:
Ah ik zie hem, dit was nogal ongelukkig qua naamgeving, aangepast. Het doet gelukkig wel wat het moet doen. FYI, in de originele code werd `StringResolver.substVars(input, session, appConstants)` gedaan, waar de 2e en 3e parameter staan voor:

```
 * @param props1 First property object in which to find substitutions
 * @param props2 Second property object in which to find substitutions
```
Zoals ik dat lees worden dus niet alleen AppConstants maar ook de session values gebruikt in substitute vars scenario.